File: backend/src/main/java/com/progressive/banking/moneytransfer/service/AuthService.java
package com.progressive.banking.moneytransfer.service;

import com.progressive.banking.moneytransfer.domain.dto.LoginRequest;
import com.progressive.banking.moneytransfer.domain.dto.LoginResponse;
import com.progressive.banking.moneytransfer.domain.dto.SignupRequest;
import com.progressive.banking.moneytransfer.domain.dto.SignupResponse;
import com.progressive.banking.moneytransfer.domain.dto.VerifyOtpRequest;

public interface AuthService {

    /**
     * Login user
     *  - validates username/password
     *  - generates JWT
     *  - returns token + accountId
     */
    LoginResponse login(LoginRequest request);

    /**
     * Signup user
     *  - validates username/password
     *  - creates User
     *  - creates Account with ₹5000 balance
     *  - generates unique 8-digit account number
     *  - returns account number + message
     */
    SignupResponse signup(SignupRequest request);
    
    // Verify using otp
    void verifyOtp(VerifyOtpRequest request);
}

File: frontend/money-transfer-ui/src/app/core/services/auth.service.ts
import { Injectable, signal } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import { LoginRequest, LoginResponse } from '../models/auth.model';
import { environment } from '../../../environments/environment';

const TOKEN_KEY = 'mts_token';
const ACCOUNT_ID_KEY = 'mts_account_id';
const HOLDER_NAME_KEY = 'mts_holder_name';

export interface SignupRequest {
  userName: string;
  email: string;
  password: string;
}

export interface SignupResponse {
  message: string;
  accountId?: number;
}

export interface VerifyOtpRequest {
  username: string;
  otp: string;
}

@Injectable({
  providedIn: 'root',
})
export class AuthService {
  readonly isLoggedIn = signal<boolean>(this.hasToken());
  readonly holderName = signal<string | null>(localStorage.getItem(HOLDER_NAME_KEY));

  constructor(private http: HttpClient) {}

  // ================= LOGIN =================
  login(payload: LoginRequest): Observable<LoginResponse> {
    return this.http.post<LoginResponse>(`${environment.apiBaseUrl}/auth/login`, payload).pipe(
      tap((res) => {
        localStorage.setItem(TOKEN_KEY, res.token);
        localStorage.setItem(ACCOUNT_ID_KEY, String(res.accountId));
        localStorage.setItem(HOLDER_NAME_KEY, res.holderName);
        this.isLoggedIn.set(true);
        this.holderName.set(res.holderName);
      })
    );
  }

  // ================= SIGNUP =================
  signup(payload: SignupRequest): Observable<SignupResponse> {
    return this.http.post<SignupResponse>(`${environment.apiBaseUrl}/auth/signup`, payload);
  }

  // ================= VERIFY OTP =================
  verifyOtp(payload: VerifyOtpRequest): Observable<string> {
    return this.http.post(
      `${environment.apiBaseUrl}/auth/verify-otp`,
      payload,
      { responseType: 'text' }
    );
  }

  // ================= LOGOUT =================
  logout(): void {
    localStorage.removeItem(TOKEN_KEY);
    localStorage.removeItem(ACCOUNT_ID_KEY);
    localStorage.removeItem(HOLDER_NAME_KEY);
    this.isLoggedIn.set(false);
    this.holderName.set(null);
  }

  // ================= HELPERS =================
  isAuthenticated(): boolean {
    return this.hasToken();
  }

  getToken(): string | null {
    return localStorage.getItem(TOKEN_KEY);
  }

  getAccountId(): number | null {
    const raw = localStorage.getItem(ACCOUNT_ID_KEY);
    return raw ? Number(raw) : null;
  }

  getHolderName(): string | null {
    return localStorage.getItem(HOLDER_NAME_KEY);
  }

  private hasToken(): boolean {
    return !!localStorage.getItem(TOKEN_KEY);
  }
}